src/components/timeline/TrackScreenshots.js

```diff
@@ -0,0 +1,261 @@
+} from '../../reducers/profile-view';
+import { withSize, type SizeProps } from '../shared/WithSize';
+
+import type {
+  ThreadIndex,
+  Thread,
+  MarkersTableWithPayload,
+} from '../../types/profile';
+import type { ScreenshotPayload } from '../../types/markers';
+import type { Milliseconds } from '../../types/units';
+import type {
+  ExplicitConnectOptions,
+  ConnectedProps,
+} from '../../utils/connect';
+
+import { ensureExists } from '../../utils/flow';
+import './TrackScreenshots.css';
+
+type OwnProps = {|
+  +threadIndex: ThreadIndex,
+  +screenshotId: string,
+  +overlayElement?: HTMLElement,
+  ...SizeProps,
+|};
+type StateProps = {|
+  +thread: Thread,
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +screenshots: MarkersTableWithPayload<ScreenshotPayload>,
+  +threadName: string,
+  +isMakingPreviewSelection: boolean,
+  +overlayElement: HTMLElement,
+|};
+type DispatchProps = {||};
+type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+type State = {|
+  offsetX: null | number,
+  pageX: null | number,
+  containerTop: null | number,
+|};
+
+export const TRACK_HEIGHT = 50;
+const HOVER_HEIGHT = 100;
+const HOVER_MAX_WIDTH_RATIO = 1.75;
+
+class Screenshots extends PureComponent<Props, State> {
+  state = {
+    offsetX: null,
+    pageX: null,
+    containerTop: null,
+  };
+
+  findScreenshotAtMouse(offsetX: number): number | null {
+    const { width, rangeStart, rangeEnd, screenshots } = this.props;
+    const rangeLength = rangeEnd - rangeStart;
+    const mouseTime = offsetX / width * rangeLength + rangeStart;
+    if (screenshots.length === 0) {
+      return null;
+    }
+    if (mouseTime < screenshots.time[0]) {
+      // Only show a screenshot for the first time we know of.
+      return null;
+    }
+    for (let i = 0; i < screenshots.length; i++) {
+      const time = screenshots.time[i];
+      if (time >= mouseTime) {
+        return i;
+      }
+    }
+    return screenshots.length - 1;
+  }
+
+  /**
+   * This function runs through all of the screenshots, and then samples the last known
+   * screenshot, and places it on the screen, making a film strip.
+   */
+  renderScreenshotStrip() {
+    const {
+      thread,
+      width: outerContainerWidth,
+      rangeStart,
+      rangeEnd,
+      screenshots,
+    } = this.props;
+    if (screenshots.length === 0) {
+      return null;
+    }
+    const images = [];
+    const rangeLength = rangeEnd - rangeStart;
+    const imageContainerWidth = TRACK_HEIGHT * 0.75;
+    const timeToPixel = time =>
+      outerContainerWidth * (time - rangeStart) / rangeLength;
+
+    let screenshotIndex = 0;
+    for (
+      let left = timeToPixel(screenshots.time[0]);
+      left < outerContainerWidth;
+      left += imageContainerWidth
+    ) {
+      // Try to find the next screenshot to fit in, or re-use the existing one.
+      for (let i = screenshotIndex; i < screenshots.length; i++) {
+        if (timeToPixel(screenshots.time[i]) <= left) {
```

Comment:
should we compare to `left + imageContainerWidth / 2`, so that a screenshot's center (instead of its left) would match the screenshot's time ?

Reply:
I would say we should base it on time, not based on the screenshot width, that way the left-hand side represents the current time. I believe this is how film editing programs do it.

Reply:
okay, I'm not familiar with this so let's do like you say :)